server/commands/src/impls.rs

```diff
@@ -573,3 +578,242 @@ pub fn seed(ctx: &mut CommandCtx) -> anyhow::Result<()> {
+        "Only players may be affected by this command, but the provided selector includes entities"
+    )]
+    NotPlayer,
+    #[error("Already banned")]
+    NoTargets,
+}
+
+#[command(usage = "ban <targets> <reason>")]
+pub fn ban_withreason(
+    ctx: &mut CommandCtx,
+    targets: EntitySelector,
+    reason: TextArgument,
+) -> anyhow::Result<()> {
+    ban_players(ctx, targets, reason.0, false)
+}
+
+#[command(usage = "ban <targets>")]
+pub fn ban_noreason(ctx: &mut CommandCtx, targets: EntitySelector) -> anyhow::Result<()> {
+    ban_players(ctx, targets, "Banned by an operator.".to_owned(), false)
+}
+
+#[command(usage = "ban-ip <targets> <reason>")]
+pub fn banip_withreason(
+    ctx: &mut CommandCtx,
+    targets: EntitySelector,
+    reason: TextArgument,
+) -> anyhow::Result<()> {
+    ban_players(ctx, targets, reason.0, true)
+}
+
+#[command(usage = "ban-ip <targets>")]
+pub fn banip_noreason(ctx: &mut CommandCtx, targets: EntitySelector) -> anyhow::Result<()> {
+    ban_players(ctx, targets, "Banned by an operator.".to_owned(), true)
+}
+
+#[derive(Debug, Error)]
+pub enum BanIpError {
+    #[error("Not a valid IP Address.")]
+    InvalidIp,
+}
+
+#[command(usage = "ban-ip <ip> <reason>")]
+pub fn banip_withreason_ip(
+    ctx: &mut CommandCtx,
+    ip: String,
+    reason: TextArgument,
+) -> anyhow::Result<()> {
+    ban_ip(ctx, ip, reason.0)
+}
+
+#[command(usage = "ban-ip <ip>")]
+pub fn banip_noreason_ip(ctx: &mut CommandCtx, ip: String) -> anyhow::Result<()> {
+    ban_ip(ctx, ip, "IP Banned by an operator.".to_string())
+}
+
+pub fn ban_ip(ctx: &mut CommandCtx, ip: String, reason: String) -> anyhow::Result<Option<String>> {
+    let ip = IpAddr::from_str(&ip).map_err(|_| BanIpError::InvalidIp)?;
+
+    {
+        let bi_lock = ctx.game.resources.get::<WrappedBanInfo>();
+        let mut ban_info = bi_lock.write().unwrap();
+
+        ban_info.ip_bans.insert(
+            ip,
+            Ban {
+                reason: reason.clone(),
+                expires_after: None,
+            },
+        );
+    }
+
+    if let Some(mut sender_message_receiver) = ctx.world.try_get_mut::<MessageReceiver>(ctx.sender)
+    {
+        let ban_confirm = Text::from(TextValue::translate_with(
+            "commands.ban.success",
+            vec![Text::from(ip.to_string()), Text::from(reason.clone())],
+        ));
+        sender_message_receiver.send(ban_confirm);
+    }
+
+    let ent = Read::<(SocketAddr, Entity)>::query()
+        .iter(ctx.world.inner())
+        .find(|x| x.0.ip() == ip)
+        .map(|x| (*x).1);
+
+    if let Some(ent) = ent {
+        ctx.game.disconnect_and_log(
+            ent,
+            &mut ctx.world,
+            &Text::from(reason),
+            "Banned by an operator.",
+        );
+    }
+
+    Ok(None)
+}
+
+pub fn ban_players(
+    ctx: &mut CommandCtx,
+    targets: EntitySelector,
+    reason: String,
+    by_ip: bool,
+) -> anyhow::Result<Option<String>> {
+    if targets.entities.len() == 0 {
```

Comment:
All PRs in feather are run through clippy as part of the CI steps. This line fires a lint, unfortunately.
```
error: length comparison to zero
   --> server/commands/src/impls.rs:688:8
    |
688 |     if targets.entities.len() == 0 {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: using `is_empty` is clearer and more explicit: `targets.entities.is_empty()`
    |
    = note: `-D clippy::len-zero` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#len_zero
```

For future reference, make sure you run `cargo clean` and `cargo clippy`. Sometimes clippy is funky without a `clean` first.